Import previously exported json data

NVanvyve opened the pull request:
In order to solve the issue #417, I added a tab to load a JSON, as well as a way to add the data found there, in an existing (or empty) project.
I realize that this was not the best way to manage the front end. But I couldn't add the form in the options section. So I did it like that to focus on the rest.
I had to add some messages, I have already noted their translation into French in `fr/LC_MESSAGES/messages.po`

Natim commented:
This is really appalling, do you mind adding a test for the upload_form?

NVanvyve commented:
I can do that. 
You mean, test if the file is a correct json?
Because I don't see how else to test the form, especially since I didn't find a test for the other types of forms

Natim commented:
I meant adding a test to make sure that importing a given json will create the project with the expected members and bills.

NVanvyve commented:
I added a test as requested and also modified some of what I had done before. I realized that it was possible to search for a user that had been deactivated. At first, I considered that they were two different people

almet commented on ihatemoney/templates/forms.html:
```diff
@@ -89,7 +89,6 @@
     {% include "display_errors.html" %}
     {{ form.hidden_tag() }}
     {{ form.file }}
-    {{ form.csrf_token() }}
```
?

NVanvyve replied:
I had a little trouble when I had to add a new form because I don't master HTML very well. That's why I created a new tab "Upload", because despite my attempts, I couldn't put it in "Settings".
But then, when I succeeded in creating this form, I had problems reading the JSON. When I searched on the internet, I was advised to add this.
It's afterwards that I realized that it didn't bring anything, so I deleted it. Especially since it was not present anywhere else.

NVanvyve commented:
Is there anything else I can do?

almet commented:
Thank you, we just need to have someone review your changes and validate them. Let me summon @Glandos and @Natim on this :-)

Natim commented on ihatemoney/tests/tests.py:
```diff
@@ -1190,6 +1190,62 @@ def test_export(self):
+            },
+        ]
+
+        with open("json_test.json", "w+") as file:
```
I am pretty sure you can use a StringIO there:

```suggestion
    from io import StringIO
        file = StringIO(json.dumps(json_to_import))
```

Natim commented on ihatemoney/tests/tests.py:
```diff
@@ -1190,6 +1190,62 @@ def test_export(self):
+        with open("json_test.json", "w+") as file:
+            json.dump(json_to_import, file)
+
+        self.client.post("/raclette/upload_json", data={"file": "json_test.json"})
```
```suggestion
        self.client.post("/raclette/upload_json", data={"file": file})
```

Natim replied:
If this doesn't work, please use [mkstemp](https://docs.python.org/2/library/tempfile.html#tempfile.mkstemp) to make sure to create the file in a temporary folder.

Natim commented on ihatemoney/tests/tests.py:
```diff
@@ -1190,6 +1190,62 @@ def test_export(self):
+
+        for i in json_to_import:
+            for j in bills:
+                if j["what"] == i["what"]:
```
This doesn't assert that all bills have been loaded properly and that none are missing.

Natim commented on ihatemoney/utils.py:
```diff
@@ -234,3 +234,29 @@ def _eval(node):
+
+
+def parse_date(string_date):
+    date = string_date.split("-")
```
please use [dateutil parser](https://dateutil.readthedocs.io/en/stable/parser.html) for this.

Natim commented on ihatemoney/web.py:
```diff
@@ -391,6 +396,83 @@ def edit_project():
+        filename = pid + "_uploaded_bills.json"
+        form.file.data.save(filename)
+        try:
+            import_project(filename)
```
There we have the same issue where we create the file locally rather than in memory.

Glandos replied:
By the way, since we already have the data available, it should be simpler to just parse them as JSON here, handles exceptions and give the JSON dict to `import_project`. Not touching the disk seems a good idea.

Natim commented on ihatemoney/web.py:
```diff
@@ -391,6 +396,83 @@ def edit_project():
+
+def import_project(file):
+    # From json : export list of members
+    json_file = json.load(open(file))
```
if we use a file descriptor rather than a filename here we would fix the previous issue.

Glandos commented on ihatemoney/translations/fr/LC_MESSAGES/messages.po:
```diff
@@ -539,6 +539,16 @@ msgstr "A dépensé"
+msgstr "Importer"
+
+msgid "Project successfully uploaded"
+msgstr "Le projet a correctement été importé"
```
Hmmm, mon intuition me guiderait vers « Le projet a été correctement importé ». Je trouve que la formulation actuelle est bancale, mais je n'ai pas de justification…

Natim replied:
```suggestion
msgstr "Le projet a été correctement importé"
```

NVanvyve commented:
I tried this but `self.assertEqual(len(bills), 3)` fail.
Obviously I can't give a file as an argument.
I think I had the same problem before but like @Natim said I wasn't looking to see if I had the right number of bills.
``` python
        tmpdir = tempfile.mkdtemp()
        predictable_filename = 'file.json'

        # Ensure the file is read/write by the creator only
        saved_umask = os.umask(0o077)

        path = os.path.join(tmpdir, predictable_filename)
        with open(path, "w") as tmp:
            json.dump(json_to_import,tmp)

        self.client.post("/raclette/upload_json", data={"file": path})

        os.remove(path)
        os.umask(saved_umask)
        os.rmdir(tmpdir)

        project = models.Project.query.get("raclette")

        bills = project.get_pretty_bills()

        # Check if all bills has been add
        self.assertEqual(len(bills), 3)
```

Glandos commented on ihatemoney/web.py:
```diff
@@ -391,6 +396,79 @@ def edit_project():
+
+def import_project(file, project):
+    # From json : export list of members
+    json_file = json.loads(file)
```
I tried to change this with the following:
```py
json_file = json.load(file)
```
The goal was to avoid double loading from memory, since the main use case is reading data from `upload_json`.
There are other changes then.

Glandos commented on ihatemoney/web.py:
```diff
@@ -391,6 +396,79 @@ def edit_project():
+    if form.validate_on_submit():
+        file = form.file.data.stream.read()
+        try:
+            import_project(file, g.project)
```
Replace this with a simpler:
```py
import_project(form.file.data.stream, g.project)
```
And remove the previous `file` declaration

Glandos commented on ihatemoney/tests/tests.py:
```diff
@@ -1190,6 +1190,156 @@ def test_export(self):
+
+        from ihatemoney.web import import_project
+
+        import_project(json.dumps(json_to_import), project)
```
This is a test, so we don't care losing some time:
```py
        file = io.StringIO()
        json.dump(json_to_import, file)
        file.seek(0)
        import_project(file, project)
```

Glandos commented on ihatemoney/tests/tests.py:
```diff
@@ -1190,6 +1190,156 @@ def test_export(self):
+
+        from ihatemoney.web import import_project
+
+        import_project(json.dumps(json_to_import), project)
```
Same here as in `test_import_new_project`:
```py
        file = io.StringIO()
        json.dump(json_to_import, file)
        file.seek(0)
        import_project(file, project)
```

Glandos commented on ihatemoney/tests/tests.py:
```diff
@@ -1190,6 +1190,215 @@ def test_export(self):
+        json_1 = [
+            {  # wrong keys
+                "checked": False,
+                "dimensions": {"width": 5, "height": 10,},
```
I don't know why the last comma in dict fails only in Py3.6 tests… But can you remove it?

And thanks for that new test! :clap:

NVanvyve replied:
Thank you. I was just looking for where the mistake could have come from.

Natim replied:
> why the last comma in dict fails only in Py3.6 tests

It is because flake8 tests are only ran with Python 3.6
